src/python/WMCore/MicroService/Unified/MSRuleCleaner.py

```diff
@@ -50,27 +116,341 @@ def execute(self, reqStatus):
+                    continue
+                if wflow['CleanupStatus'][pline.name]:
+                    self.wfCounters['cleaned'][pline.name] += 1
+        elif wflow['RequestStatus'] == 'announced' and not wflow['ParentageResolved']:
+            # NOTE: We skip workflows which are not having 'ParentageResolved'
+            #       flag, but we still need some proper logging for these.
+            msg = "Skipping workflow: %s - 'ParentageResolved' flag set to false."
+            msg += "\nWill retry again in the next cycle."
+            self.logger.info(msg, wflow['RequestName'])
+        elif wflow['RequestStatus'] == 'announced' and wflow['ParentageResolved']:
+            for pline in self.cleanuplines:
+                try:
+                    pline.run(wflow)
+                except Exception as ex:
+                    msg = "%s: General error from pipeline. Error: \n%s. "
+                    msg += "\nWill retry again in the next cycle."
+                    self.logger.exception(msg, pline.name, str(ex))
+                    continue
+                if wflow['CleanupStatus'][pline.name]:
+                    self.wfCounters['cleaned'][pline.name] += 1
+        else:
+            # We shouldn't be here:
+            msg = "Skipping workflow: %s - "
+            msg += "Does not fall under any of the defined categories."
+            self.logger.info(msg, wflow['RequestName'])
+
+        # Archive:
+        try:
+            self.plineArchive.run(wflow)
+            self.wfCounters['archived'] += 1
+        except MSRuleCleanerArchival as ex:
+            msg = "%s: Archival Error: %s. "
+            msg += "\nWill retry again in the next cycle."
+            self.logger.error(msg, wflow['PlineMarkers'][-1], ex.message())
+        except Exception as ex:
+            msg = "%s General error from pipeline. Error: \n%s. "
+            msg += "\nWill retry again in the next cycle."
+            self.logger.exception(msg, wflow['PlineMarkers'][-1], str(ex))
+        return wflow
+
+    def setPlineMarker(self, wflow, pName=None):
+        """
+        A function intended to mark which is the pipeline currently working
+        on the workflow. It is supposed to be called always as a first function
+        in the pipeline.
+        :param  wflow:   A MSRuleCleaner workflow representation
+        :param  pName:   The name of the functional pipeline
+        :return wflow:
+        """
+        # NOTE: The current functional pipeline MUST always be appended at the
+        #       end of the 'PlineMarkers' list
+
+        # First get rid of the default:
+        if not wflow['PlineMarkers']:
+            wflow['PlineMarkers'] = []
+
+        # Then push our current value into the markers list:
+        wflow['PlineMarkers'].append(pName)
+
+        # Populate the list of flags to be used later:
+        if pName not in wflow['RulesToClean']:
+            if pName in [pline.name for pline in self.cleanuplines]:
```

Comment:
Just an observation (same for the block below):
```
[pline.name for pline in self.cleanuplines]
```
is an static code, even though it takes some CPU cycles for every workflow and pileup to be executed. Likely something that's worth changing.

Reply:
I am open for suggestions. This was the shortest, cleanest and cheapest (as it comes to CPU cycles and lines of code) I could come up with. And I also tested all the use cases I could cover in the same format (as for the check a line bellow) so that it is crystal clear just from reading the code which are the use cases covered in these checks.

Reply:
What I had in mind was something like
```
self.cleanupPipeNames = [pline.name for pline in self.cleanuplines]
```

somewhere upstream, like `__init__` or `execute`. Then here it would be just 
```
if pName in self.cleanupPipeNames:
```

Python interpreter might be smart enough to optimize the original line though, making no use of the enhancement I'm suggesting. If you feel like not changing this code anymore as well, that's fine either.

Reply:
Done